util/flask_util.py

```diff
@@ -16,4 +22,125 @@ def problem(type, status, title, detail=None, instance=None, headers={}):
+        :param private: If this is True, then the response contains
+            information from an authenticated client and should not be stored
+            in intermediate caches.
+        """
+        max_age = max_age or 0
+        try:
+            max_age = int(max_age)
+        except ValueError:
+            max_age = 0
+        self.max_age = max_age
+        if private is None:
+            if self.max_age == 0:
+                # The most common reason for max_age to be set to 0 is that a resource
+                # is _also_ private.
+                private = True
+            else:
+                private = False
+        self.private = private
+
+        body = response
+        if isinstance(body, etree._Element):
+            body = etree.tostring(body)
+        elif not isinstance(body, (bytes, unicode)):
+            body = unicode(body)
+
+        super(Response, self).__init__(
+            response=body,
+            status=status,
+            headers=self._headers(headers or {}),
+            mimetype=mimetype,
+            content_type=content_type,
+            direct_passthrough=direct_passthrough
+        )
+
+    def __unicode__(self):
+        """This object can be treated as a string, e.g. in tests.
+
+        :return: The entity-body portion of the response.
+        """
+        return self.data
+
+    def _headers(self, headers={}):
+        """Build an appropriate set of HTTP response headers."""
+        # Don't modify the underlying dictionary; it came from somewhere else.
+        headers = dict(headers)
+
+        # Set Cache-Control based on max-age.
+        if self.private:
+            private = "private"
+        else:
+            private = "public"
+        if self.max_age and isinstance(self.max_age, int):
+            # A CDN should hold on to the cached representation only half
+            # as long as the end-user.
+            client_cache = self.max_age
+            cdn_cache = self.max_age / 2
+            cache_control = "%s, no-transform, max-age=%d, s-maxage=%d" % (
+                private, client_cache, cdn_cache
+            )
+
+            # Explicitly set Expires based on max-age; some clients need this.
+            expires_at = datetime.datetime.utcnow() + datetime.timedelta(
+                seconds=self.max_age
+            )
+            headers['Expires'] = format_date_time(
+                time.mktime(expires_at.timetuple())
+            )
+        else:
+            # Missing, invalid or zero max-age means don't cache at all.
+            cache_control = "%s, no-cache" % private
+        headers['Cache-Control'] = cache_control
+
+        return headers
+
+
+class OPDSFeedResponse(Response):
+    """A convenience specialization of Response for typical OPDS feeds."""
+    def __init__(self, response=None, status=None, headers=None, mimetype=None,
+                 content_type=None, direct_passthrough=False, max_age=None,
+                 private=None):
```

Comment:
This method is really verbose, compared to OPDSEntryResponse, because I wanted to make it really easy for anything that used to create a Response object to create an OPDSFeedResponse instead, whether it was using positional arguments or keyword arguments. This isn't as necessary for OPDSEntryResponse, which is only used in a couple places, so I wrote the simpler version for that class that requires keyword arguments.